packages/wrangler/src/publish.ts

```diff
@@ -89,27 +89,61 @@ export default async function publish(props: Props): Promise<void> {
         "You cannot configure [text_blobs] with an ES module worker. Instead, import the file directly in your code, and optionally configure `[build.upload.rules]` in your wrangler.toml"
       );
     }
-
-    const { modules, resolvedEntryPointPath, bundleType } = await bundleWorker(
-      props.entry,
-      destination.path,
-      {
+    const { modules, resolvedEntryPointPath, bundleType, exports } =
+      await bundleWorker(props.entry, destination.path, {
         serveAssetsFromWorker: props.experimentalPublic,
         jsxFactory,
         jsxFragment,
         rules: props.rules,
         tsconfig: props.tsconfig ?? config.tsconfig,
+      });
+
+    // Some validation of durable objects + migrations
+    if (config.durable_objects.bindings.length > 0) {
+      // intrinsic [durable_objects] implies [migrations]
+      if (
+        config.durable_objects.bindings.filter(
+          (binding) => !binding.script_name
+        ).length > 0 &&
+        config.migrations.length === 0
+      ) {
+        console.warn(
+          "In wrangler.toml, you have configured intrinsic [durable objects], but no [migrations]. This may not work as expected until you add a [migrations] section to your wrangler.toml. Refer to https://developers.cloudflare.com/workers/learning/using-durable-objects/#durable-object-migrations-in-wranglertoml for more details."
+        );
+      }
+
+      // Make sure the worker exports all the declared durable objects
+      for (const binding of config.durable_objects.bindings) {
+        const missingExports = [];
+        if (!binding.script_name && !exports.includes(binding.class_name)) {
+          missingExports.push(binding.class_name);
+        }
+        if (missingExports.length > 0) {
+          throw new Error(
+            `The worker is missing exports for durable object bindings: ${missingExports.join(
```

Comment:
Oh huh I forgot to add tests for this one. Adding.

Reply:
Done, discovered a couple more issues when doing so, filed https://github.com/cloudflare/wrangler2/issues/731